internal/backends/python/python.go

```diff
@@ -311,6 +304,167 @@ func makePythonPoetryBackend(python string) api.LanguageBackend {
 	}
 }
 
+var pythonGuessRegexps = util.Regexps([]string{
+	// The (?:.|\\\n) subexpression allows us to
+	// match match multiple lines if
+	// backslash-escapes are used on the newlines.
+	`from (?:.|\\\n) import`,
+	`import ((?:.|\\\n)*) as`,
+	`import ((?:.|\\\n)*)`,
+})
+
+// makePythonPipBackend returns a backend for invoking poetry, given an arg0 for invoking Python
+// (either a full path or just a name like "python3") to use when invoking Python.
+func makePythonPipBackend(python string) api.LanguageBackend {
+	var pipFlags []PipFlag
+
+	b := api.LanguageBackend{
+		Name:                 "python3-pip",
+		Specfile:             "requirements.txt",
+		Alias:                "python-python3-pip",
+		FilenamePatterns:     []string{"*.py"},
+		Quirks:               api.QuirksAddRemoveAlsoInstalls | api.QuirksNotReproducible,
+		NormalizePackageName: normalizePackageName,
+		GetPackageDir: func() string {
+			// Check if we're already inside an activated
+			// virtualenv. If so, just use it.
+			if venv := os.Getenv("VIRTUAL_ENV"); venv != "" {
+				return venv
+			}
+
+			if outputB, err := util.GetCmdOutputFallible([]string{
+				"python",
+				"-c", "import site; print(site.USER_SITE)",
+			}); err == nil {
+				return string(outputB)
+			}
+
+			return ""
+		},
+		SortPackages: pkg.SortPrefixSuffix(normalizePackageName),
+
+		Search: func(query string) []api.PkgInfo {
+			results, err := SearchPypi(query)
+			if err != nil {
+				util.Die("failed to search pypi: %s", err.Error())
+			}
+			return results
+		},
+		Info: info,
+		Add: func(ctx context.Context, pkgs map[api.PkgName]api.PkgSpec, projectName string) {
+			//nolint:ineffassign,wastedassign,staticcheck
+			span, ctx := tracer.StartSpanFromContext(ctx, "pip install")
+			defer span.Finish()
+
+			cmd := []string{"pip", "install"}
+			for _, flag := range pipFlags {
+				cmd = append(cmd, string(flag))
+			}
+			for name, spec := range pkgs {
+				name := string(name)
+				spec := string(spec)
+
+				cmd = append(cmd, name+spec)
+			}
+			// Run install
+			util.RunCmd(cmd)
+			// Determine what was actually installed
+			{
+				outputB, err := util.GetCmdOutputFallible([]string{
+					"pip", "freeze",
+				})
```

Comment:
here's one of my gripes about upm not having defined semantics. technically we're locking packages here! which means pip is, in fact, reproducible! it just so happens that the spec file(s) can also be lock file(s)... maybe we shouldn't be marking pip as non-reproducible? and maybe for the `Lock` func we can run `pip freeze` and replace all the lines without a constraint with the constrained version...

Reply:
from the lens of a behavioral descriptivist (given we haven't prescribed *any* behavior for upm at all, lmao), i still think we should have `Lock` for the pip backend. upm's behavior for `upm lock` has been to allow the package manager to ensure that all dependencies are locked. if dependencies aren't locked, then they should be

Reply:
I'm happy to go along with this, save for where to put the secondary `requirements.txt` (and how to tie it in).

We have some choices:
1. One, singular `requirements.txt` that has all the locked transitive deps. Now the user has transitive deps that need to be solved when either bumping or uninstalling libraries.
2. One `requirements.txt` that has the `upm add`'d deps, another `requirements.lock.txt` (or `.full.txt` or whatever), which gets used by `upm install`, but that breaks user expectations if switching between `upm` and `pip` (Do I `pip install -r requirements.txt` or `pip install -r requirements.lock.txt`?)
3. `requirements.txt` _is_ the `pip freeze` file, and `upm add` adds to/removes from `requirements.upm.txt` or similar

Given the ambiguity and hassle for the user, I would actually prefer to leave locking in the hands of the user until https://discuss.python.org/t/what-is-the-status-on-a-pip-lock-file/6669/8 .

Benefits to deferring:
- Those that know/care about reproducibility can choose to lock in their deployments.
- Those that are using docker to build anyway get dependency locking inherent to their production images.
- Removing the burden of maintaining the transitive deps keeps transitive deps unpinned, guiding users towards a glorious future of maintainability for their projects instead of "Nothing works! I need to delete my requirements.txt" (which was very close to what we saw with `replit-py` the other week)

Reply:
gotcha. then we probably shouldn't lock the deps we add at all right? since that's more likely to cause breakages in the package?

Reply:
I think the particulars of the package are fine, but transitive deps are not OK, same as in many other languages. If something 5 layers down has a CVE, we should permit that update to make its way through naturally.

The benefit is giving some stability in which library API we're targeting, so if we `upm add` right before a major version bump our code doesn't start breaking immediately after we create the project.